### ISSUE TITLE
Unable to compress cubes with negative samples

### ISSUE THREAD

**Issue by SindreOsnes** [equinor/seismic-zfp#36]
If a segy cube has a negative sample the np_numpy_float to bytes failes when converting to a signed integer.

Further the fact that the samples are treated as integers can cause a problem later as the sample points are not necessarily correct. E.g. a small non zero offset from a "unit multiple" step wil cause the actual sample values and the sgz sample values to deviate periodically.  If the samples are not to be recalculated based on the headers is zfp, they should instead be stored as signed floats, `'<f'` instead of `'<I'`.

**Comment by da-wad** [comment]
Good catch! I think the two issues need solving a little separately... the possibility of negative sample times is from the original SEGY standard, but non-integer-microsecond sample intervals was introduced in SEGY rev2.

**1) Negative sample times:**

"Delay recording time", bytes 109-110, is the responsible trace header. It's a 16 bit integer (in ms, not us), and must be signed if negative sample times are observed? As this _must_ be an integer, probably all that needs doing is (re-)creating an np_float_to_bytes_signed function (using ```struct.pack("<i")```) and putting what segyio tells you in there. Although for completeness you may possibly need to check the scalar in bytes 215–216 is correctly applied...


**2) Non-integer-microsecond sample intervals**

SEGY rev2 had to give SeismicZFP some surprises... rev0 and rev1 _only_ support sample intervals of 2-byte integers of microseconds, and rev2 made non-integers possible in an obscure way.

See:
 - Bytes 3217-3218: https://seg.org/Portals/0/SEG/News%20and%20Resources/Technical%20Standards/seg_y_rev1.pdf
 - Bytes 3273-3280 https://seg.org/Portals/0/SEG/News%20and%20Resources/Technical%20Standards/seg_y_rev2_0-mar2017.pdf

The ```zslices_list``` generation may be updated to check whether bytes 3273-3280 are non-zero in the binary header (which SeismicZFP keeps a copy of and can read via segyio), before defaulting to using what is stored as a uint32 in the SGZ header.

**Comment by fmell** [comment]
I have had a look at this issue now and have a possible fix for your first point (negative sample times) in https://github.com/equinor/seismic-zfp/commit/a9801e39ddc4f8f6a3ccf831aea10c6851275d0b .

This seems to work for SEGY files with a negative Delay recording time (Trace header bytes 109-110), as well as files written with previous versions of SeismicZFP. Bytes 215-216 does not seem to be handled by segyio, so I assume it will be okay to ignore this for now.

On your second point it seems that segyio does not have support for non-integer sample intervals yet. Since we rely on segyio for reading this I assume this needs to be implemented there first before we can support it.

**Comment by SindreOsnes** [comment]
@fmell Thah sounds a little bit wierid to me as if you open seismic cube, using segyio, that I provided to you the samples array is clearly floats. The float in the samples array seems to come from these lines:

- [https://github.com/equinor/segyio/blob/3044598e6e5dbaa2aff540dfcb6481a1c3c0d242/python/segyio/segyio.cpp#L1280](https://github.com/equinor/segyio/blob/3044598e6e5dbaa2aff540dfcb6481a1c3c0d242/python/segyio/segyio.cpp#L1280)
- [https://github.com/equinor/segyio/blob/96c7ca251c1bc5cee718be360de97b485bf02a15/python/segyio/open.py#L175](https://github.com/equinor/segyio/blob/96c7ca251c1bc5cee718be360de97b485bf02a15/python/segyio/open.py#L175)

These lines imply that if the sample interval (i.e.) the int is not an integer multiple of 1000 we end up with non integer sample intervals in the sample objects. e.g. 4875 -> 4.875. Further it implies that the samples array is a float array no matter what due to the divide bby 1000.0 constraint in the second entry. As such seismic zfp should copy that logic to calculate the samples themselves in order to proved accurate access patterns.

**Comment by da-wad** [comment]
Thanks for taking this on @fmell  Unfortunately for everyone, the very minimum weirdness of SEG-Y is "a little bit"....

Digging deeper into the [C code behind segyio](https://github.com/equinor/segyio/blob/96c7ca251c1bc5cee718be360de97b485bf02a15/lib/src/segy.c#L1305), you'll see that the sample interval is fetched out of the SEG-Y file as an integer number of microseconds and converted to a float before the code @SindreOsnes cites. Therefore I can't see how there would be any numeric discrepancies?

segyio _does_ deliver this as a numpy array of floats though (float64, in fact on my machine), so we definitely have an issue with types. seismic-zfp does [a 'python division' by 1000](https://github.com/equinor/seismic-zfp/blob/cf6e563fdc136a054bd1bc2cfbab6225ad523fde/seismic_zfp/read.py#L214) to get the sample_rate in ms. This means that if it's a multiple of 1000 you'll get a list of ints, and a list of floats if not. Explicit is better than implicit... says PEP20 :-/

The quick fix may be to change that line to say `/= 1000.0`

Then additionally change from lists to numpy arrays...

**Comment by fmell** [comment]
I have done some more digging now, both into the standard ([SEG-Y_r2.0: SEG-Y revision 2.0 Data Exchange format](https://seg.org/Portals/0/SEG/News%20and%20Resources/Technical%20Standards/seg_y_rev2_0-mar2017.pdf)), and the code of segyio and SeismicZFP.

It seems like there are several ways to store the Sample Interval in a SEGY file.

**Option 1 - Binary File Header bytes 3217-3218**
```
Sample interval. Microseconds (µs) for time data, Hertz (Hz) for frequency
data, meters (m) or feet (ft) for depth data. 
```

**Option 2 - Standard Trace Header bytes 117-118**
```
Sample interval for this trace. Microseconds (µs) for time data, Hertz (Hz) for
frequency data, meters (m) or feet (ft) for depth data.
If the fixed length trace flag in bytes 3503–3504 of the Binary File Header is set,
the sample interval in every trace in the SEG-Y file is assumed to be the same
as the value recorded in the Binary File Header and this field is ignored. If the
fixed length trace flag is not set, the sample interval may vary from trace to trace
```

**Option 3 - Binary File Header bytes 3273-3280**
```
Extended sample interval, IEEE double precision (64-bit). If nonzero, this
overrides the sample interval in bytes 3217–3218 with the same units.
```

Option 1 and 2 are defined as two-byte two's complement **or** unsigned integers. I assume sample intervals can't be negative, so I think these should actually be unsigned integers. segyio treats these as signed integers, but there are complaints in equinor/segyio#494 where someone tried to load a 40 ms sampled cube and segyio returns a negative sample interval.

Option 3 is a floating point number. Does not seem to be supported by segyio.

What segyio does:
- gets option 2 from the first trace
- also gets the Delay recording time (Standard Trace Header bytes 109-110) from the first trace
- if both option 1 and 2 are set with different values, or if neither is set, it falls back to a hard coded sample interval of 4ms
- if only one is set or both is the same it will use that to generate the samples array which is in milliseconds.
- segyio does not seem to care about the Scalar in bytes 215-216.

What SeismicZFP does:
- length of the samples array from segyio are put this in Seismic ZFP Header bytes 4-7 as uint32
- first element of the samples array from segyio are put in Seismic ZFP Header bytes 16-19 as uint32
- binary File Header bytes 3217-3218 are put in bytes 28-31 as uint32

**Comment by da-wad** [comment]
Good digging, this agrees with what I've seen. So... given:

- (a) The fix in [a9801e3](https://github.com/equinor/seismic-zfp/commit/a9801e39ddc4f8f6a3ccf831aea10c6851275d0b) which enables a negative first element of the samples array
and
- (b) Converting the SeismicZFP samples list to a numpy float array

... the only time SeismicZFP would give a different samples array to segyio is if Option 1 is missing, or disagrees with Option 2.

This could be remedied by taking the segyio sample interval, converting to microseconds and storing as uint32?

Until and unless Option 3 is supported by segyio, this would be exact as far as I can see.
